calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detectors/chessboard_detector.py

```diff
@@ -53,16 +63,45 @@ def detect(self, img: np.array, stamp: float):
 
             resized_detection = self.resized_detection.value
             resized_max_resolution = self.resized_max_resolution.value
+            local_padding = self.padding.value
+            max_lost_frames = self.max_lost_frames.value
+
+        def get_roi(corners, frame_shape, padding=local_padding):
+            # Region to keep track of the chessboard in the next frame
+            x_min, y_min = np.min(corners, axis=0).ravel().astype(int) - padding
+            x_max, y_max = np.max(corners, axis=0).ravel().astype(int) + padding
+            x_min, y_min = max(0, x_min), max(0, y_min)
+            x_max, y_max = min(frame_shape[1], x_max), min(frame_shape[0], y_max)
+            return (x_min, y_min, x_max, y_max)
 
         h, w = img.shape[0:2]
         grayscale = to_grayscale(img)
-
         if not resized_detection or max(h, w) <= resized_max_resolution:
-            (ok, corners) = cv2.findChessboardCorners(grayscale, (cols, rows), flags=flags)
-
-            if not ok:
-                self.detection_results_signal.emit(img, None, stamp)
-                return
+            if self.roi is None or self.lost_frames >= max_lost_frames:
+                (detected, corners) = cv2.findChessboardCorners(
+                    grayscale, (cols, rows), flags=flags
+                )
+                # if chessboard was found, keep track of the region to try to detect easily in the next frame
+                if detected:
+                    self.roi = get_roi(corners, img.shape[:2])
+                    self.lost_frames = 0
```

Comment:
`self.lost_frames` is reset in a protected way now, but it is not here

Reply:
In:
`self.lost_frames = self.max_lost_frames.value`

`self.max_lost_frames` is shared with the parameter edit widget, so it needs to be protected.

`restart_lost_frames_counter` modifies `self.lost_frames` in another thread, and as such it needs to be protected.

To avoid a flurry of locks, you can load one time into a local variable, and write (using a lock) just once after you finish whatever processing was being done.

Another choice is to use async signals with Qt instead of using a method (this is the best approach according to this framework).

If you want to understand how things go under the hood you can print the thread ids